### ISSUE TITLE
Parsing of "Plain format" strings in Waypoint Calculator

### ISSUE THREAD

**Issue by S-Bartfast** [cgeo/cgeo#9791]
Greetings all.

I have been much time away from this project but now finally with some time I hope to reacquaint myself with you all again. As the original designer of the Waypoint Calculator and finally with some time to spare I am hopeful that I may be able to contribute to things again. After some reading of the current issues it appears that many of the issues appear to be related to the parsing of the "Plain format" text in the calculator. Is this the case?

As it was, the original intent of the calculator was to enter coordinates in one of the regular formats (such as `DDD°MM.MMM'`) then substitute individual digits with equations (such as `N 51°24.AB5'`, with `A=a, B=a+2`). The "Plain format" was really only added as an option as an after thought more for completeness sake than anything, but I take it this "Plain" format is being used most frequently (such as `"N 51 24.a(a+2)5"`). Is this the case?

Assuming the "Plain" format is indeed being used frequently I take it there are usability issues regarding the parsing of the text. If so could we discuss the short comings of the current parser here and I will do my best to address them.

Thank you,
Slarti.

**Comment by S-Bartfast** [comment]
Hmm, I seem to be having trouble working out how to navigate this service again: 😊 
How do I attach labels to this issue (ie. WP Calculator), and how do I assign it to myself?

**Comment by S-Bartfast** [comment]
The first issue I've come across seems to be related to white space in the text: #7910

From memory this was done intentionally and uses the same rules for parsing as dose the regular "Plain format" parser for the non-calculator input. That is `"42 12.345"` would be parsed as `N 42­° 12.345'`, and `"42 12 345"` would be parsed as `N 42° 12' 345` and `"42 12.3 4 5"` would be parsed as an error.

From memory, the way the calculator works is that it substitutes all the equations for their respective values then attempts to parse the result as a valid coordinate format. Given that the result is required to have specific white space requirements in order to be parsed as a valid coordinate the formula entered into the field is required to have the same.

**Comment by murggel** [comment]
@S-Bartfast 
I am a "normal" user of the app and a c++ developer who is new to java and app development. In the last time I used the calculator a lot - most plain format - and stumbled over some issues. So I took a look into it. Mainly to parse the personal note to extract formula for a waypoint.
See my attempts in #9633 / #9211.

Please have a look on that.
Still missing there to complete the workflow, but I am working on it, is to extract the evaluation of the formula without the Calculator-dialog.

I am stuck on extracting the other formats.
So if you have a different approach, suggestions, we can discuss it...

Regards
murggel

**Comment by moving-bits** [comment]
Hi @S-Bartfast, welcome back!

I've adjusted assignment and label for this issue.

If you'd like to filter for open issues having the label "WP Calc" you can use the following link:
https://github.com/cgeo/cgeo/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%22WP+Calc%22+

**Comment by moving-bits** [comment]
> and uses the same rules for parsing as dose the regular "Plain format" parser for the non-calculator input.

I remember that we've made some changes to that pattern in the context of parsing waypoints from personal notes to reduce the amount of false positives. Maybe we need to use separate patterns for regular parsing in text and parsing in the calculator for plain format.

**Comment by S-Bartfast** [comment]
Glad to see this feature/project still has such strong support: 🙂

@murggel
> I am a "normal" user of the app and a c++ developer who is new to java and app development

Huh, that's exactly where I was when I started working on this calculator... about 10 years of C++ development and about 10 days of Java 😉. Fortunately, in the intervening years I have since learnt much about Java app development and would now consider myself "almost competent": 😁

**Comment by S-Bartfast** [comment]
@moving-bits 
> I've adjusted assignment and label for this issue.

Thank you. How does one do this? I was trying to do it myself but couldn't workout how. Does one need elevated privileges or anything?

**Comment by moving-bits** [comment]
> > I've adjusted assignment and label for this issue.
> Thank you. How does one do this? I was trying to do it myself but couldn't workout how. Are elevated privileges required or anything?

It can be done using the little symbols in the "Assignees" and "Labels" sections right beside the issues pane:

![grafik](https://user-images.githubusercontent.com/3754370/104450870-11b37000-55a1-11eb-8f9e-75800b12fac1.png)

(But maybe some special priviledges are necessary.)

**Comment by S-Bartfast** [comment]
Hmm, I don't seem to have the gear/option icon there myself:
![image](https://user-images.githubusercontent.com/6573522/104452283-e65b7000-55f6-11eb-8dd3-26dd45df4a3d.png)
(and I accidentally unassigned myself while looking: 🤦‍♂️)

**Comment by fm-sys** [comment]
AFAIK only users which are part of the c:geo development team can change labels and stuff like that...

**Comment by geocachermgo** [comment]
Hello S-Bartfast,
 I'd like to have the possiblity to
convert more digit numbers like 12345 to 12.345 and to use faction like 12345/1000

You may have a look at https://www.guballa.de/plausibaer
where there is such a possibility..

**Comment by S-Bartfast** [comment]
Hello geocachermgo. Are you able to give an example (preferably in English) of a cache which demonstrates your problem.

As discussed in #6699 [Multi digit solutions in WP Calc do not support decimals] (which I see you also commented in) the calculator intentionally only works with integer solutions. The main reason for this is because it's just not clear how decimals should be handled.

If we can think of a consistent way to handle decimal expansions in the general case I'm more than happy to implement it but coming up with a general solution is surprisingly difficult; believe me, I spent much time thinking it over when I designed the calculator in the first place: 😉

**Comment by geocachermgo** [comment]
I went through the last riddles I found, none of them in english, because our trip to England in Summer was canceled. This are my last where the result was spanning over the decimal point.
GC8P022 the east value is given as Ost 9 0(J+160) and J is bigger than 1000.
GC8NYV9 east value: Ost: 9 0(E – 4xF + A)
GC3EABQ N 48 (Y + 6807) / 1000, E 8 (X - 9345) / 1000
Had you had a look to the plausibaer examples? He uses xx.xxx as special marker for a value spanning over the decimal point.

**Comment by fm-sys** [comment]
GC8NZR5 is also such an example.

The calculated number let's say is `12345` and must be interpreted as N 49° `12.345'`

**Comment by S-Bartfast** [comment]
Hmm, I'm finding it very difficult to understand what you're requesting.

GC8P022 and GC8NYV9 don't display anything meaningful to me. Is a premium membership required perhaps?

GC8NYV9 does at least return a cache but exactly what it's asking is non the less lost on me.

Could you possibly post a screenshot of what you are seeing potentially with annotations as to how you would like it to be different.

> Had you had a look to the plausibaer examples?
I followed the link (https://www.guballa.de/plausibaer) but it seemed to go to an online tool rather than a specific example. I did attempt to past in the example text and it plots three points on a map but I wasn't sure what to do with it.


As far as I'm aware the calculator works as intended and it's not clear to me how it should behave differently.

**Comment by S-Bartfast** [comment]
> GC8NZR5 is also such an example.
> 
> The calculated number let's say is `12345` and must be interpreted as N 49° `12.345'`

Okay, think I have this one:
![image](https://user-images.githubusercontent.com/6573522/104725571-5fd59880-5786-11eb-8ad0-6497dbe49187.png)

Are you able to give me the values for variables **a** through **j**?

**Comment by fm-sys** [comment]
Amazing! This is indeed currently possible. I didn't  know that the WP computer is such a powerful tool...

But two general questions about the WP Calc:
- Why do we need to distinguish between upper and lowercase letters? 
- Why do the inputs overlap to the right? Smartphone users are used to scrolling vertically, while horizontal scrolling is quite awkward...
What do you think? Can we improve these two things?

**Comment by S-Bartfast** [comment]
When the calculator was conceived it was expected that letters would be used to replace digits and that the letters would be expressed as expressions like this:
![image](https://user-images.githubusercontent.com/6573522/104744710-4f7ee700-57a1-11eb-89de-df918548f135.png)

It then became apparent that the same values were often used in multiple expressions so "free variables" were added like this:
![image](https://user-images.githubusercontent.com/6573522/104745394-29a61200-57a2-11eb-85d2-9cfa56c6aa10.png)


Note you can tilt the calculator into a landscape orientation to see both the "expressions" and "variables" simultaneously if you wish:
![image](https://user-images.githubusercontent.com/6573522/104745430-332f7a00-57a2-11eb-8fd5-9df0e90b9c86.png)


The upper and lower case letters were intended to help distinguish between "expressions" and "variables" and eliminate circular dependencies.


It was only after this that the "Plain" format was introduced and although it is handy I had misgivings about including it as I feared it may complicate things too much. There was none the less desire for the "Plane" format though so I implemented it, more for completeness sake than anything. It now seems this "Plain" format is the most discussed format and unfortunately it is less intuitive to use than the others.

**Comment by S-Bartfast** [comment]
> * Why do we need to distinguish between upper and lowercase letters?

The reason for this is that _Equations_  (uppercase letters) can use _Variables_ (lowercase letters). In order to eliminate circular dependencies _equations_ are not able to use other _equations_, and _variables_ must only consist of arithmetic expressions.

> * Why do the inputs overlap to the right? Smartphone users are used to scrolling vertically, while horizontal scrolling is quite awkward...

I thought a lot about this and adopted the horizontal sliding panel in an attempt to keep the "variables" with the "equations". I know it seems strange at first but once you start to use it it does become fairly intuitive... and since then "swiping right" has only become a more natural gesture 😉

>   What do you think? Can we improve these two things?

I'm certainly open to suggestions. As I say, I actually put a lot of thought into the current layout but that doesn't necessarily mean it's the best it can be. I'm more than happy to consider other options if anyone can think of anything that would improve things. 🙂

**Comment by geocachermgo** [comment]
Since there is a button to convert the copy of the formula  to lower case, I'm fine with that.

**Comment by fm-sys** [comment]
> In order to eliminate circular dependencies equations are not able to use other equations, and variables must only consist of arithmetic expressions.

Even if it would be a bit coding effort it would be kind of cool to detect and block such circular dependencies so that no differentiation between equations and variables is needed.

On the other hand I understand the longer I think about why you did it that way and not differently. It has advantages! 

However, if we leave it like it is, we should limit the variables input to numbers and  `+-*/%` and convert uppercase to lowercase letters automatically in the equations fileds. (Which is currently possible via the button). However, this conversion should be done via an `onTextChangeListener` so that you are still be able to paste uppercase letters in their which will then be converted automatically instead of limiting the input characters directly

**Comment by geocachermgo** [comment]
> Hmm, I'm finding it very difficult to understand what you're requesting.
Sorry for that.
> GC8P022 and GC8NYV9 don't display anything meaningful to me. Is a premium membership required perhaps?
Yes you are right, sorry, I didin't checked that beforehand.

> Could you possibly post a screenshot of what you are seeing potentially with annotations as to how you would like it to be different.
> 
It's already possible to cover the 3 digit decimal by one letter and to cover the 2 digits of the integral minutes as one number,
I would like to have the 5 digits of the mm.mmm as one variable.
Without using the (A/1000) .  (A%1000) workaround

> > Had you had a look to the plausibaer examples?
> > I followed the link (https://www.guballa.de/plausibaer) but it seemed to go to an online tool rather than a specific example. I did attempt to past in the example text and it plots three points on a map but I wasn't sure what to do with it.
Is a tool to calculate not one but an arbitrary number of possible solutions and shown them one a map.
It very useful if you do not have the complete solution and the formula has some restrictions, and e.g. you can then check the places in a forest that are close to a path, or in a city the places that are not a road or builing.
 
The examples are given here: 
https://translate.google.com/translate?sl=de&tl=en&u=https://www.guballa.de/plausibaer-beispiel-1
https://translate.google.com/translate?sl=de&tl=en&u=https://www.guballa.de/plausibaer-beispiel-2
https://translate.google.com/translate?sl=de&tl=en&u=https://www.guballa.de/plausibaer-beispiel-3
 
It also has a two (+1) stage approach (here from example 2  (=beispiel-2):
First you have the (lower case) variables. eg N 47 33.y E 7 55.x
then you have the "equations" eg. 
p = a + r + o
y = p - 133
x = p + 384
and the third step is the define the value or possibility for the "interim variables" eg by
fixed value o = 7
or enumeration o = 1,2,8
or range o = 7..12

Why I mentioned it here, is because there is the possibility to deliberately spread one variable to several digits (see example-3) by using quotation marks e.g.
N 48 "xx.xxx" E 6 "5y.yyy" 
with x=23456 and y=8765 this would result in N 48 23.456 E 6 58.765

**Comment by fm-sys** [comment]
> Why I mentioned it here, is because there is the possibility to deliberately spread one variable to several digits (see example-3) by using quotation marks e.g.
N 48 "xx.xxx" E 6 "5y.yyy"
with x=23456 and y=8765 this would result in N 48 23.456 E 6 58.765

That sounds like a good way how we could implement this also in c:geo...

**Comment by geocachermgo** [comment]
![IMG_20210115_205124](https://user-images.githubusercontent.com/43813636/104772285-7a524680-5773-11eb-983b-6b1eb27aa283.jpg)
One option would be to extent the logic "if there is space right of a varible extend it to the right.
So here it should be resolved to 48 51.234

**Comment by geocachermgo** [comment]
![Uploading IMG_20210115_205740.jpg…]()

as it is done here within the decimals

**Comment by S-Bartfast** [comment]
> Since there is a button to convert the copy of the formula to lower case, I'm fine with that.

Oh, I didn't even realise there was a button.
I had thought input was restricted to only lowercase letters but I see something has changed. I'm definitely not in favour of allowing uppercase letters to be added to equations. Was there any reason this was changed? Was it perhaps an oversite?

In any event it definitely seems the wrong way to go to me.

**Comment by S-Bartfast** [comment]
> However, if we leave it like it is, we should limit the variables input to numbers and `+-*/%` and convert uppercase to lowercase letters automatically in the equations fileds. (Which is currently possible via the button). However, this conversion should be done via an `onTextChangeListener` so that you are still be able to paste uppercase letters in their which will then be converted automatically instead of limiting the input characters directly

Definitely! I probably didn't know about TextChangeListeners at the time I wrote this (I was very new to Android development) but yes, this is clearly the way it should be done: 👍

**Comment by S-Bartfast** [comment]
> Without using the (A/1000) . (A%1000) workaround

I definitely agree, the workaround is a bit of a hack and I'm not much a fan of it either. I just didn't know what else to do.

As said in the other thread I did spend a lot of time banging my head against how to handle decimal arithmetic and ultimately opted to deal exclusively with integers. Part of the problem is where to put padding zeros. As in, in MM.MMM' format should `12.34` be expanded to `12.340'` or `12.034'`. I know 12.340' is mathematically the correct answer but padding with leading zeros (even after the decimal point) seems to be the norm in geocaching. There were also other issues that arose so I just opted to deal exclusively with integer arithmetic.

Ultimately I ended up opting for formats that most closely followed the other means of entering coordinates, those being simple text input of wither `DDD.DDDDD°`, `DDD° MM.MMM'` or `DDD° MM' SS.SSS"`... and of course the dreaded "Plain" format: 😱

**Comment by S-Bartfast** [comment]
> there is the possibility to deliberately spread one variable to several digits (see example-3) by using quotation marks e.g.
> N 48 "xx.xxx" E 6 "5y.yyy"
> with x=23456 and y=8765 this would result in N 48 23.456 E 6 58.765

Yes, it would be a possibility to adopt that as a syntax but it's definitely not the way things are handled at the moment. I tried to shy away from too much "intelligent" parsing of input and rather just opted for simple letter substitutions.

This is the kind of hornets nest I was afraid of when making the "Plain" format available. The other formats with the individual buttons representing digits did away with the kind of intelligent parsing your discussing here. Clearly it was a lot more restrictive but ultimately it was a lot simpler for both developers and users alike.

Once we start parsing "free text" we firstly need the legal formats to be understood by both the developers as well as users (which was hard enough for simply typing numbers in plain format). Once the complexity of performing calculation substitutions to the format it became significantly more difficult and now we are talking about defining a syntax for expressing additional formatting. This is of course all possible but we may well be looking into something much bigger and deeper than any of us expect.

**Comment by moving-bits** [comment]
> Oh, I didn't even realise there was a button.

The button is relatively new. I introduced it not long ago because many listings use upper case letters, and it's quite a hassle to change those on a mobile's display - didn't know of `onTextChangeListener`...
